framework/src/org/checkerframework/common/jaifinference/JaifInferenceUtils.java

```diff
@@ -0,0 +1,485 @@
+
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Attribute.Array;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.TypeSymbol;
+import com.sun.tools.javac.code.Symbol.VarSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Type.ArrayType;
+import com.sun.tools.javac.code.Type.ClassType;
+/**
+ * The purpose of this class is to allow a whole-program type inference with
+ * the aid of .jaif files.
+ *
+ * This class has public methods that writes into .jaif files updating
+ * field types and method return types of classes according to their ATM.
+ * The types are always updated to be the LUB between the previous type and
+ * the current type.
+ *
+ * @author pbsf
+ *
+ */
+public class JaifInferenceUtils {
+
+    /**
+     * Path to where .jaif files will be written to and read from.
+     * This path is relative to where the CF's javac command is executed.
+     */
+    public final static String JAIF_FILES_PATH = "build/jaif-files/";
+
+    /**
+     * List of annotations that are ignored - not written into .jaif
+     * files.
+     */
+    private static List<AnnotationMirror> annotationsToIgnore =
+            new ArrayList<AnnotationMirror>();
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param methodElt is the method element.
+     * @param atf is the annotated type factory.
+     * @return the ATM of methodElt in the .jaif file of the class with symbol
+     * classSymbol, or null if it can't be found.
+     */
+    public static AnnotatedTypeMirror getMethodReturnTypeInJaif(
+            ClassSymbol classSymbol, ExecutableElement methodElt,
+            AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return null; // Static block.
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, new AScene());
+            if (clazz == null) return null;
+            String methodName = JVMNames.getJVMMethodName(methodElt);
+            AMethod method = clazz.methods.get(methodName);
+            if (method != null && method.returnType != null) {
+                Set<Annotation> prevAnnos = method.returnType.tlAnnotationsHere;
+                return setOfAnnotationsToATM(prevAnnos, atf, false,
+                            methodElt.getReturnType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * @param classSymbol is the symbol representing the class containing the method.
+     * @param fieldNode is the node representing the field.
+     * @param atf is the annotated type factory.
+     * @return the ATM of fieldNode in the .jaif file of the class with symbol
```

Comment:
Here and above, I'm not sure what it means for a dataflow node to be in the .jaif file.